### PR TITLE
BUSINESS RULE: "R - 00824" and "R - 00818"has been refactored using rule validator Interface!!

### PR DESCRIPTION
**Pull request by AVandana-OSI**
Rule - 00824

Set REFERRAL CLIENT.Disposition Code = 'A' for (all) REFERRAL_CLIENT (where REFERRAL_CLIENT > REFERRAL.Received_Date minus REFERRAL_CLIENT > CLIENT.Birth_Date is less than 19 years) when REFERRAL.Referral_Response_Type = 'Evaluate Out' and REFERRAL.Approval_Status_Type = 'Approved'.

Rule - 00818

If the Response Decision is 'Evaluate Out', then Rationale is mandatory and Agency Referred to is enabled and mandatory, else Agency Referred To is disabled.

### CODE REVIEW

File: src/main/java/gov/ca/cwds/rest/business/rules/R00818SetReferredResourceType.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+   */
+  public R00818SetReferredResourceType(ScreeningToReferral screeningToReferral) {
+    super();
+    this.screeningToReferral = screeningToReferral;
+  }
+
+  @Override
+  public boolean isValid() {
+    Boolean referredToResourceType = Boolean.FALSE;
+    if (screeningToReferral.getResponseTime() == 1519) {
```
Comment (tparkerosi) [inline src/main/java/gov/ca/cwds/rest/business/rules/R00818SetReferredResourceType.java]: 1559 is a "Magic Number". Better to Replace it with a Constant that names what that number is. If it is an LOV, we can use the LOV.

File: src/main/java/gov/ca/cwds/rest/business/rules/R00824SetDispositionCode.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    super();
+    this.screeningToReferral = screeningToReferral;
+    this.incomingParticipant = incomingParticipant;
+  }
+
+  @Override
+  public boolean isValid() {
+    return clientAge() < 19 && screeningToReferral.getResponseTime() == 1519
+        && screeningToReferral.getApprovalStatus() == 122;
+  }
```
Comment (tparkerosi) [inline src/main/java/gov/ca/cwds/rest/business/rules/R00824SetDispositionCode.java]: 122 && 1559 is a "Magic Number". Better to Replace it with a Constant that names what that number is. If it is an LOV, we can use the LOV. 19 can be replaced with 'adult' or something similar.

File: src/main/java/gov/ca/cwds/rest/services/ParticipantService.java [hunk 3]
```diff
@@ -260,9 +262,11 @@ private ReferralClient processReferralClient(ScreeningToReferral screeningToRefe
       String referralId, MessageBuilder messageBuilder, Participant incomingParticipant,
       String clientId) {
 
+    RuleValidatator ruleValidator =
+        new R00824SetDispositionCode(screeningToReferral, incomingParticipant);
     ReferralClient referralClient =
         ReferralClient.createWithDefault(ParticipantValidator.selfReported(incomingParticipant),
-            incomingParticipant.isClientStaffPersonAdded(), dispositionCode(screeningToReferral),
+            incomingParticipant.isClientStaffPersonAdded(), ruleValidator.isValid() ? "A" : "",
```
Comment (tparkerosi) [inline src/main/java/gov/ca/cwds/rest/services/ParticipantService.java]: We may want to pull out the ruleValidator from the constructor. It is a bit harder to read and debugging would be difficult

File: src/main/java/gov/ca/cwds/rest/services/cms/ReferralService.java [hunk 4]
```diff
@@ -274,7 +275,8 @@ public String createCmsReferralFromScreening(ScreeningToReferral screeningToRefe
         drmsAllegationDescriptionDoc, drmsErReferralDoc, drmsInvestigationDoc,
         screeningToReferral.isFiledWithLawEnforcement(), screeningToReferral.isFamilyAwareness(),
         govEnt, screeningToReferral.getName(), dateStarted, timeStarted,
-        screeningToReferral.getResponseTime(), referredToResourceType(screeningToReferral),
+        screeningToReferral.getResponseTime(),
```
Comment (tparkerosi) [inline src/main/java/gov/ca/cwds/rest/services/cms/ReferralService.java]: 3325 is also a 'magic number'. We should name it with a variable/constant

File: src/test/java/gov/ca/cwds/rest/business/rules/doctool/R06224DontAllowBlanksInReferralStartDateAndTimeTest.java [hunk 5]
```diff
@@ -396,7 +396,7 @@ public void testForStartedAtEmptyFailure() throws Exception {
     } catch (BusinessValidationException e) {
       Set<IssueDetails> issues = e.getValidationDetailsList();
       // TODO: fix. Changed from 9-11-7
-      assertThat(issues.size(), is(equalTo(7)));
+      assertThat(issues.size(), is(equalTo(8)));
```
Comment (tparkerosi) [inline src/test/java/gov/ca/cwds/rest/business/rules/doctool/R06224DontAllowBlanksInReferralStartDateAndTimeTest.java]: I have no idea what 7 or 8 is or why that number is significant. This should be a variable that help identify the significance of the number

Comment (AVandana-OSI) [reply src/test/java/gov/ca/cwds/rest/business/rules/doctool/R06224DontAllowBlanksInReferralStartDateAndTimeTest.java]: Its a test case

**Pull request merged**